### ISSUE TITLE
seek of closed file

### ISSUE THREAD

**Issue by v-python** [python-pillow/Pillow#1630]
I wanted to calculate a histogram of a multi-page TIFF file.

``` python
        im = Image.open( fn )
        colors = im.getcolors( im.width * im.height )
        if args.hist:
            for cnt, col in colors:
                allcolors[ col ] += cnt
            for iz in range( 1, im.n_frames ):
                im.seek( iz )
                colors = im.getcolors( im.width * im.height )
                for cnt, col in colors:
                    allcolors[ col ] += cnt
```

The above code would fail on the seek, with "seek of closed file" error.  I tried first to add the extra open operation below just before the seek.  It helped some, but not in every case.

``` python
        im = Image.open( fn )
        colors = im.getcolors( im.width * im.height )
        if args.hist:
            for cnt, col in colors:
                allcolors[ col ] += cnt
            for iz in range( 1, im.n_frames ):
                im = Image.open( fn ) # does getcolors implicitly close????
                                      # without the open, get "seek of closed
                                      # file" error on line below.
                im.seek( iz )
                colors = im.getcolors( im.width * im.height )
                for cnt, col in colors:
                    allcolors[ col ] += cnt
```

Randomly (or maybe experience based) guessing, I then tried hoisting the reference to n_frames out of the loop, as below.  That seemed to make it reliable across a number of test cases that failed otherwise.

``` python
        im = Image.open( fn )
        imgcnt = im.n_frames
        colors = im.getcolors( im.width * im.height )
        if args.hist:
            for cnt, col in colors:
                allcolors[ col ] += cnt
            for iz in range( 1, imgcnt ):
                im = Image.open( fn ) # does getcolors implicitly close????
                                      # without the open, get "seek of closed
                                      # file" error on line below.
                im.seek( iz )
                colors = im.getcolors( im.width * im.height )
                for cnt, col in colors:
                    allcolors[ col ] += cnt
```

With this solution in hand, I tried removing the extra open inside the loop, but some failures recurred when I did so.  So this last code is my "production solution" for now, but it is not at all clear why the initial code attempt at the top produces errors.

Pillow 3.0.0, Python 3.5.0, Windows 10.

**Comment by radarhere** [comment]
Testing with the test image `multipage.tiff`, I can't replicate. Any chance that you could point towards an image that does replicate the problem?

**Comment by v-python** [comment]
Can I send one to you privately?  One file that most reliably produced errors is a © file from a book project.  But the code above is wrapped in a loop:  for fn in filelist:   When the filelist contained one item, failures were less frequent. When the filelist contained multiple items, the failures were more frequent.  I sort of forgot that in my report above.  So it might be possible that adding the loop and a few copies of multipage.tiff that the problem would show up.  If I can find multipage.tiff I'll try that with my full code.

**Comment by radarhere** [comment]
You can find multipage.tiff at https://github.com/python-pillow/Pillow/blob/master/Tests/images/multipage.tiff?raw=true

If it's a copyright image, I'd personally prefer to leave it alone. Hopefully the issue can be replicated some other way.

**Comment by v-python** [comment]
I found it, it is tiny. The test I suggested did not fail with my full code with workarounds, nor without the workarounds.   Let me see if I can tweak the image and remove the © content.

**Comment by v-python** [comment]
OK, in hopes that it has to do with the image size or compression type, I simply selected all the material and blacked it out, and that still produces the failure.  But dragging it in here, GitHub says "Unfortunately, we don't support that file type."  How can I submit the this modified file?

**Comment by v-python** [comment]
(I'm fairly inexperienced with GitHub.)

**Comment by v-python** [comment]
And if it hadn't been for GitHub's offer to attach files by dropping them, I'd have thought to just put it on my web site in the first place...  http://nevcal.com/temporary/m1.tif

**Comment by radarhere** [comment]
As a start, I can report that when I test with that file, I find that the following is the minimal code required to replicate -

    from PIL import Image
    im = Image.open('m1.tif')
    for iz in range(1, im.n_frames):
        im.seek(iz)
        im.load()

This line - https://github.com/python-pillow/Pillow/blob/master/PIL/TiffImagePlugin.py#L1053 - is closing the file pointer, causing the error that is raised later. This is because at that point, the __next variable has a value of 0, presumably believing there to be no next frame.

**Comment by v-python** [comment]
Thanks for your analysis. Always good when it can be confirmed on a different machine, by a different person.

**Comment by therockmandolinist** [comment]
Are there any updates on this? I'm getting issues with `ValueError: seek of closed file` as well when looping over `img.n_frames`. I could just open the image again and loop over that, I guess... doesn't seem like I should have to,  though.

**Comment by sekrause** [comment]
This is the same issue as #2049.

**Comment by radarhere** [comment]
This looks like it was fixed by #2330
